internal/server/kubeapi_handler.go

```diff
@@ -53,48 +63,69 @@ func (s *Server) KubeapiHandler(rw http.ResponseWriter, req *http.Request) {
 		rw.WriteHeader(http.StatusInternalServerError)
 		return
 	}
-	log.Debugf("[%s] REQ OK t=%s %+v", tunnelID, timeout, req)
+	log.Debugf("[%s] REQ OK t=%s target: %s, req: %+v", tunnelID, timeout, target.String(), req)
 
-	client, cfg, err := s.GetClientFromKubeconfig(tunnelID, timeout)
+	client, err := s.GetClientFromKubeconfig(tunnelID, timeout)
 	if err != nil {
 		log.Errorf("Error getting client for tunnel %s: %s", tunnelID, err)
 		rw.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
 	// Create proxy and set the transport to remotedialer client
-	proxy := proxy.NewUpgradeAwareHandler(target, client.Transport, false, false, er)
+	proxy := httputil.NewSingleHostReverseProxy(target)
 
-	upgradeTransport, err := makeUpgradeTransport(cfg, client.Transport)
-	if err != nil {
-		return
+	// Wrap the transport with LoggingTransport
+	proxy.Transport = &LoggingTransport{
+		Transport: client.Transport, // Use the existing transport
 	}
-	proxy.UpgradeTransport = upgradeTransport
 
-	req.URL.Scheme = target.Scheme
-	req.Host = target.Host
-	req.URL.Path = target.Path
-	log.Debugf("[%s] REQ DONE: %+v", tunnelID, req)
+	// Modify the Director function to change the request
+	originalDirector := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		originalDirector(req)
+		req.URL.Scheme = target.Scheme
+		req.Host = target.Host
+		req.URL.Path = target.Path
+
+		// Preserve the Upgrade header for HTTP/1.1 requests
+		if req.ProtoMajor == 1 {
+			if upgrade := req.Header.Get("Upgrade"); upgrade != "" {
+				log.Debugf("[%s] Preserving Upgrade header: %s", tunnelID, upgrade)
+			}
+		} else {
+			// Remove the Upgrade header for HTTP/2 requests
+			req.Header.Del("Upgrade")
```

Comment:
Can I ask you @gcgirish  to elaborate a bit more to help me understand why http/1.1 upgrade header is preserved and http/2 not?
And why http/2 transport is disabled later in code here https://github.com/open-edge-platform/cluster-connect-gateway/pull/34/files#diff-ed971d5c02a4d3f62fe17fedb49921905dbcce4805ed593964a0a0b2120f691bR162?
Is http/2 update request dropped to not support SPDY?

Reply:
http/2 does not support upgrade of connections to SPDY or websockets when requested by K8s API server to stream some connections that require it (like exec, cp, attach, logs kind of operations with kubectl). That is why we try to disable it here and the code references you mentioned.